placement/algo/helper.go

```diff
@@ -0,0 +1,373 @@
+
+	ph.buildHostHeap(hostCapacityAscending)
+	return ph
+}
+
+func (ph *placementHelper) buildRackToHostMap() {
+	for _, h := range ph.hostShards {
+		if _, exist := ph.rackToHostsMap[h.Host().Rack()]; !exist {
+			ph.rackToHostsMap[h.Host().Rack()] = make(map[placement.HostShards]struct{})
+		}
+		ph.rackToHostsMap[h.Host().Rack()][h] = struct{}{}
+		for _, shard := range h.Shards() {
+			ph.assignShardToHost(shard, h)
+		}
+	}
+}
+
+func (ph *placementHelper) buildHostHeap(hostCapacityAscending bool) {
+	overSizedRack := 0
+	overSizedHosts := 0
+	rackSizeMap := make(map[string]int)
+	for rack := range ph.rackToHostsMap {
+		rackHostNumber := len(ph.rackToHostsMap[rack])
+		if float64(rackHostNumber)/float64(ph.getHostLen()) >= 1.0/float64(ph.rf) {
+			overSizedRack++
+			overSizedHosts += rackHostNumber
+		}
+		rackSizeMap[rack] = rackHostNumber
+	}
+
+	targetLoadMap := ph.buildTargetLoadMap(rackSizeMap, overSizedRack, overSizedHosts)
+	ph.targetLoad = targetLoadMap
+	ph.hostHeap = newHostHeap(ph.hostShards, hostCapacityAscending, targetLoadMap, ph.rackToHostsMap)
+}
+
+func (ph *placementHelper) buildTargetLoadMap(rackSizeMap map[string]int, overSizedRackLen, overSizedHostLen int) map[string]int {
+	targetLoad := make(map[string]int)
+	for _, host := range ph.hostShards {
+		rackSize := rackSizeMap[host.Host().Rack()]
+		if float64(rackSize)/float64(ph.getHostLen()) >= 1.0/float64(ph.rf) {
+			// if the host is on a over-sized rack, the target load is topped at shardLen / rackSize
+			targetLoad[host.Host().ID()] = int(math.Ceil(float64(ph.getShardLen()) / float64(rackSize)))
+		} else {
+			// if the host is on a normal rack, get the target load with aware of other over-sized rack
+			targetLoad[host.Host().ID()] = ph.getShardLen() * (ph.rf - overSizedRackLen) / (ph.getHostLen() - overSizedHostLen)
+		}
+	}
+	return targetLoad
+}
+
+func (ph placementHelper) getHostLen() int {
+	return len(ph.hostShards)
+}
+
+func (ph placementHelper) getShardLen() int {
+	return len(ph.uniqueShards)
+}
+
+func (ph placementHelper) canAssignRack(shard uint32, from placement.HostShards, toRack string) bool {
+	if from != nil {
+		if from.Host().Rack() == toRack {
+			return true
+		}
+	}
+	for host := range ph.shardToHostMap[shard] {
+		if host.Host().Rack() == toRack {
+			return false
+		}
+	}
+	return true
+}
+
+func (ph placementHelper) canAssignHost(shard uint32, from, to placement.HostShards) bool {
+	if to.ContainsShard(shard) {
+		return false
+	}
+	return ph.canAssignRack(shard, from, to.Host().Rack())
+}
+
+func (ph placementHelper) assignShardToHost(shard uint32, to placement.HostShards) {
+	to.AddShard(shard)
+
+	if _, exist := ph.shardToHostMap[shard]; !exist {
+		ph.shardToHostMap[shard] = make(map[placement.HostShards]struct{})
+	}
+	ph.shardToHostMap[shard][to] = struct{}{}
+}
+
+func (ph placementHelper) removeShardFromHost(shard uint32, from placement.HostShards) {
+	from.RemoveShard(shard)
+
+	delete(ph.shardToHostMap[shard], from)
```

Comment:
nope, the updated hostShard can be in or out of the ph.hostShards, assign or remove a shard does not change that.